### ISSUE TITLE
Questions on a standalone config

### ISSUE THREAD

**Issue by happysalada** [lucaslorentz/caddy-docker-proxy#165]
Thanks for a great repo! I have a couple of questions on a standalone installation of caddy with docker swarm.

What should the base Caddyfile contain?
- it contains config.example.com, with a respond Hello world. Is it safe to assume it can be skipped? (it seems everything can be safely generated) What is unclear to me is that all of the tls are labeled as "internal", should there be some that are not to generate the external facing ssl certificate? I tried adding a new domain name with no pre-existing ssl certificate, when trying to access, I get an ssl error. When verifying inside the caddy docker image, I have a file at the corresponding domain
/data/caddy/certificates/local/union.blackstonefutures.co.za

Let me know if this is not the right place to post this

Here is the url in question
https://union.blackstonefutures.co.za/

And here is the relevant docker label part

```
      labels:
        caddy: www.$DOMAIN_NAME union.blackstonefutures.co.za
        caddy.reverse_proxy: "{{upstreams 80}}"
        caddy.tls: "internal"
```

the normal domain is `www.forunion.com/` and works fine, however the new domain does not.

**Comment by francislavoie** [comment]
Could you be more specific about how your config looks? It's hard to know if we're speaking about the same thing unless you're more specific about what you tried.

**Comment by happysalada** [comment]
let me know if you need more details

**Comment by francislavoie** [comment]
`tls internal` is meant to tell Caddy to generate self-signed certificates. It sound like this is not what you want, since you're trying to make your site publicly accessible?

You should be able to remove `caddy.tls: "internal"` altogether and Caddy will attempt to get a certificate issued by Let's Encrypt.

You can read about Automatic HTTPS here: https://caddyserver.com/docs/automatic-https

**Comment by happysalada** [comment]
I tried, without the caddy-basic-content part in the docker file and it doesn't work.
The problem is that it cannot be empty either, docker doesn't allow that. Is the default to just leave it with config.example.com? (Just want to make sure, it seems a bit unintuitive)

When I keep tls: "internal", one of my url works (forunion.com) but not the other (union.blackstonefutures.co.za), it's little weird to me.

Trying to remove `caddy.tls: "internal"` I get failures
`union_caddy.1.irk5ilfda6ei@Ubuntu-1804-bionic-64-minimal    | [union.blackstonefutures.co.za] acme: error: 403 :: urn:ietf:params:acme:error:unauthorized :: Invalid response from https://union.blackstonefutures.co.za/.well-known/acme-challenge/y5QG77QkqHqRW9roSagsa3Zu9MqtNFosdR_R9WElgqY [2606:4700:3037::681b:b0c1]: "<!DOCTYPE html>\n<!--[if lt IE 7]> <html class=\"no-js ie6 oldie\" lang=\"en-US\"> <![endif]-->\n<!--[if IE 7]>    <html class=\"no-js ", url:`
My server is accessible on port 80 though  (in this case you can see it's responding with a 404)
Do I have to add a redirect rule for the acme challenge to caddy?

**Comment by francislavoie** [comment]
Is your site fronted by cloudflare or something?

**Comment by happysalada** [comment]
yes, I realize now that must be the issue. Also I have an automated http to https redirect on cloudflare.
After checking a bit, I'm guessing.
- there should be no http to https redirect on cloudflare (since the redirect is taken care of by caddy from what I understand)
- The cloudflare ssl setting should be set to full (Encrypts end-to-end, using a self signed certificate on the server), or full-strict (Encrypts end-to-end, but requires a trusted CA or Cloudflare Origin CA certificate on the server)

(the weird thing is, forunion.com and union.blackstonefurures.co.za are both fronted by cloudflare, maybe different options have been enabled.)

**Comment by happysalada** [comment]
Removing automated redirect from http to https on cloudflare, the SSL option set as full, and the acme staging server, I was able to get some certificates and one of my domains is working. I'm investigating the ssl options on the other domain to see why it's not working.

All the example showed the intenal option enabled, a "real life" example would help a lot. (with a commented out staging acme server as well)
something like

```
        caddy: mydomain.com
        caddy.reverse_proxy: "{{upstreams 80}}"
        # comment out the following line when you've made sure your config works
        caddy.tls.ca: https://acme-staging-v02.api.letsencrypt.org/directory 
```

currently my server only exposes port 80, but I'm having a look at exposing port 443 as well, since the caddy default challenge is the ALPN one. (I guess this is the best practice)

**Comment by francislavoie** [comment]
Caddy actually randomly picks between HTTP or ALPN and if either fails it tries the other. This is so that if one method is taken down or removed by LE for whatever reason (has happened before with the no longer supported SNI challenge), Caddy will be robust.

Sorry I don't have any more specifics to offer regarding cloudflare, I don't use it.

One option could be to use the DNS challenge instead which will work even when behind cloudflare. That involves compiling Caddy with an additional plugin to provide support for whichever DNS provider you use.

https://caddy.community/t/how-to-use-dns-provider-modules-in-caddy-2/8148

**Comment by whitestrake** [comment]
Browsing to the erroring URL, it looks like Cloudflare is fronting your site with an Error 526 (Invalid SSL certificate). This explains the HTML document returned when the `/.well-known` URI was requested.

This requires you to have the site set on **Strict** SSL mode; that error status does not appear otherwise (see: https://developers.cloudflare.com/ssl/origin/ssl-modes/).

According to LetsEncrypt's documentation, HTTP-01 challenges occur over port 80, but accept redirects to HTTPS (see: https://letsencrypt.org/docs/challenge-types/#http-01-challenge).

Caddy itself does not issue HTTP->S redirects for ACME challenges (the challenge is "absorbed" by certificate automation logic if it goes straight to Caddy). The conclusion drawn is that Cloudflare issued a HTTP->S redirect to the initial HTTP-01 challenge (as you mention you had this configured).

This MUST NOT happen if the site is also set to **Strict**, or Cloudflare will shoot itself in its own foot and then refuse to serve your site like a petulant child.

<img width="492" alt="Screen Shot 2020-07-16 at 1 52 20 pm" src="https://user-images.githubusercontent.com/9500326/87624646-9432b280-c76b-11ea-8904-1d1b662cdab5.png">

Theoretically, EITHER reverting from **Strict** to **Full** OR disabling Cloudflare-side redirects will solve this initial issue, and subsequent renewals shouldn't run into the same problem for as long as those renewals take place with a still-valid certificate (this should be 99.99999% of renewals as Caddy attempts renewals with plenty of time to spare).

My recommendation is to allow Caddy to handle HTTP->S redirects in all cases, anyway, as the benefit from having Cloudflare do this is speed (since CF can issue the redirect from any POP rather than waiting for your origin), but this only matters for the first request (the one the client makes before moving to HTTPS thereafter).

**Comment by happysalada** [comment]
Thanks for the taking the time to take a look.
There are two options on cloudflare for HTTPS
- automated HTTPS rewrites
- force HTTPS
Once disabling both, it all works fine.
(maybe it's worth adding a note in the docs, or perhaps it out of scope).


So one question remains.
- What is the envisioned use case for the base Caddyfile?
I currently have it with the default config.example.com because I can't have it empty and I can't put anything else, because everything is taken care of by the docker labels?
(just thinking of how to improve potentially the docs.)

**Comment by lucaslorentz** [comment]
@happysalada 

The use case is to do any configuration that doesn't belong to a specific docker service/container.
For example, proxies to external servers, storage configuration, global options and so on.

It is an optional feature. It is not required to have a base caddyfile.
You can just remove everything related to it.

A PR improving the README would be very welcome.
As a non native English speaker, I struggle a bit there.

**Comment by happysalada** [comment]
I added comments to the yaml file directly, let me know if you think that is inappropriate
https://github.com/lucaslorentz/caddy-docker-proxy/pull/166

**Comment by happysalada** [comment]
closed by #166
